docs/proposals/08-shoot-additional-container-runtimes.md

```diff
@@ -0,0 +1,139 @@
+
+The motivation behind this proposal is to make all of this functionality accessible to clusters managed by Gardener.
+
+### Goals
+
+* Gardener must default to configuring its managed clusters with the CRI interface instead of the legacy Dockershim. This change must be transparent to Gardener users and must simply work out-of-the-box for new and existing clusters.
+* Low-level runtimes like gVisor or Kata Containers are provided as gardener extensions which are (optionally) installed into a landscape by the Gardener operator. There must be no runtime-specific knowledge in the core Gardener code.
+* It shall be possible to configure multiple low-level runtimes in Gardener clusters, on the Worker Group level.
+
+### Non-Goals
+
+* Exposing the configuration of the underlying runtimes to the user is beyond the scope of this proposal. The initial version will support a default common-sense configuration for each of the underlying runtimes. If required, future enhancements will make it possible to fine-tune the underlying runtimes.
+
+## Proposal
+
+Gardener today assumes that all supported operating systems have Docker pre-installed in the base image. Starting with Docker Engine 1.11, Docker itself was [refactored](https://www.docker.com/blog/docker-engine-1-11-runc/) and cleaned-up to be based on the [containerd](https://containerd.io/) library. The first phase would be to change the default Kubelet configuration as described [here](https://kubernetes.io/docs/setup/production-environment/container-runtimes/#containerd) so that Kubernetes would use containerd instead of the default Dockershim. This will be implemented for CoreOS, Ubuntu, and SuSE-JeOS. 
+
+Once Gardener clusters are migrated to use CRI, we will implement two Gardener extensions, providing gVisor and Kata Containers as options for Gardener landscapes. The `WorkerGroup` specification will be extended to allow specifying a list of additional required Runtimes for nodes in that group. For example:
+```yaml
+workers:
+- name: worker-b8jg5
+  machineType: m5.large
+  volumeType: gp2
+  volumeSize: 50Gi
+  autoScalerMin: 1
+  autoScalerMax: 2
+  maxSurge: 1
+  containerRuntimes:
+  - gvisor
+  - kata-containers
+  machineImage:
+    name: coreos
+    version: 2135.6.0
+```
+
+Each extension will need to address the following concern:
+
+1. Add the low-level runtime binaries to the worker nodes. Each extension should store the runtime binary in a way that is accessible to a node to download and install during startup.
```

Comment:
Where are these binaries coming from? We probably should add them to the container image of the extension itself, or have a separate container image that we can download and use to copy them over to the nodes' file systems. See also https://github.com/gardener/gardener/pull/1771#issuecomment-569601529 for the motivation (private cloud/on-prem deployments). /cc @ccwienk @vlerenc

Reply:
I am not convinced, we can abstract OS from container runtime. In some cases, the OS anyways mandates the container runtime (RHEL). In any case, it requires very specific handling. If we invent contracts like these, we may end up extending the contract in multiple ways, almost never elegantly, so maybe we should bite the bullet and rather say, OS and container runtime is one thing/one extension. WDYT?

Reply:
Our latest attempt at sorting this shows that if the underlying Kubelet is using containerd, it's possible to install the required binaries from a privileged container - maybe from a `DaemonSet` running in the cluster - and reload the containerd without messing up the kubelet.

We'll update the design to show that the new runtime extension type is basically a ValidatingWebhook coupled with a helm chart to install the correct daemonset into the shoot. The binaries will be provided in a Docker image referenced in the chart - which we will need to maintain as part of the delivery of the specific extension extension.

Reply:
> and reload the containerd without messing up the kubelet.

Do you mean "restarting" by "reloading" here? If so, (how) is this possible without killing all containers? Or is "reloading" something else? Or is this only for the initial provisioning and not possible without a rolling update of the nodes?

What about @vlerenc's question from above? You say there is not(hing/ much) OS-specific as we just need to copy some binaries to the file system and "reload" the containerd?

Reply:
Example implementation for Ubuntu and gVisor:
https://github.com/nimrodoron/gardener-extensions/pull/10/files

Reply:
The default "docker-like" engine always exists, additional runtimes like gVisor are optional addons.